// @aurora/core-engine/src/api/ws.ts
// File: packages/core-engine/src/api/ws.ts
// Purpose: WebSocket server for real-time events

import { WebSocketServer, WebSocket } from 'ws';
import { Server as HttpServer } from 'http';
import { WSMessage } from '@aurora/shared';
import { logger } from '../utils/logger';

export class WSEventServer {
  private wss: WebSocketServer;
  private clients: Set<WebSocket> = new Set();

  constructor(server: HttpServer) {
    this.wss = new WebSocketServer({ server, path: '/v1/logs' });

    this.wss.on('connection', (ws: WebSocket) => {
      this.clients.add(ws);
      logger.info(`WebSocket client connected. Total clients: ${this.clients.size}`);

      ws.on('close', () => {
        this.clients.delete(ws);
        logger.info(`WebSocket client disconnected. Total clients: ${this.clients.size}`);
      });

      ws.on('error', (error) => {
        logger.error('WebSocket error:', error);
      });

      // Send welcome message
      this.sendToClient(ws, {
        type: 'connection.heartbeat',
        payload: { message: 'Connected to Aurora Engine' },
        timestamp: Date.now(),
      });
    });

    logger.info('WebSocket server configured at /v1/logs');
  }

  // Send message to specific client
  private sendToClient(client: WebSocket, message: WSMessage) {
    if (client.readyState === WebSocket.OPEN) {
      client.send(JSON.stringify(message));
    }
  }

  // Broadcast to all connected clients
  broadcast(message: WSMessage) {
    const payload = JSON.stringify(message);
    this.clients.forEach((client) => {
      if (client.readyState === WebSocket.OPEN) {
        client.send(payload);
      }
    });
  }

  // Graceful shutdown
  close() {
    this.clients.forEach((client) => client.close());
    this.wss.close();
    logger.info('WebSocket server closed');
  }
}
