// @aurora/shared/src/types.ts
// Core event types for the system

export interface TranscriptEvent {
  id: string;
  timestamp: number;
  text: string;
  partial: boolean;
}

export interface IntentDecision {
  intent: string;
  confidence: number;
  slots: Record<string, string>;
  requiresApproval: boolean;
  timestamp: number;
}

export interface ActionRequest {
  action: string;
  params: Record<string, any>;
  requestId: string;
  timestamp: number;
}

export interface ActionResult {
  ok: boolean;
  data?: any;
  error?: string;
  code?: number;
  requestId: string;
  timestamp: number;
}

export interface Recipe {
  id: string;
  name: string;
  description?: string;
  actions: ActionRequest[];
  createdAt: number;
  updatedAt: number;
}

export interface RecipeRunResult {
  recipeId: string;
  results: ActionResult[];
  ok: boolean;
}

// WebSocket message types
export type WSMessageType = 
  | 'stt.partial'
  | 'stt.final'
  | 'intent.preview'
  | 'action.result'
  | 'heartbeat';

export interface WSMessage<T = any> {
  type: WSMessageType;
  payload: T;
  timestamp: number;
}

// API Error type
export interface APIError {
  error: string;
  code: number;
}

// Settings
export interface EngineSettings {
  port: number;
  token: string;
  modelPath: string;
  autoStart: boolean;
}
