// @aurora/core-engine/src/utils/config.ts
// File: packages/core-engine/src/utils/config.ts
// Purpose: Load and validate environment variables

import { config as dotenvConfig } from 'dotenv';
import path from 'path';

// Load .env file
dotenvConfig({ path: path.resolve(__dirname, '../../.env') });

interface Config {
  port: number;
  token: string;
  nodeEnv: 'development' | 'production';
  modelDir: string;
  autostartSTT: boolean;
  logLevel: string;
}

function getConfig(): Config {
  const requiredVars = ['ENGINE_PORT', 'ENGINE_TOKEN'];
  
  // Validate required variables
  for (const varName of requiredVars) {
    if (!process.env[varName]) {
      throw new Error(`Missing required environment variable: ${varName}`);
    }
  }

  return {
    port: parseInt(process.env.ENGINE_PORT || '8080', 10),
    token: process.env.ENGINE_TOKEN!,
    nodeEnv: (process.env.NODE_ENV as any) || 'development',
    modelDir: process.env.MODEL_DIR || './models/vosk-model-small-en-us-0.15',
    autostartSTT: process.env.AUTOSTART_STT === 'true',
    logLevel: process.env.LOG_LEVEL || 'info',
  };
}

export const config = getConfig();
