// @aurora/web-dashboard/src/lib/useWebSocket.ts
import { useEffect, useRef, useState, useCallback } from 'react';
import { WSMessage } from '@aurora/shared';

const WS_URL = import.meta.env.VITE_WS_URL || 'ws://localhost:8080';
const MOCK_MODE = import.meta.env.VITE_MOCK_MODE === 'true';

interface UseWebSocketOptions {
  onMessage?: (message: WSMessage) => void;
  reconnectInterval?: number;
  maxReconnectAttempts?: number;
}

export function useWebSocket(options: UseWebSocketOptions = {}) {
  const {
    onMessage,
    reconnectInterval = 5000,
    maxReconnectAttempts = 10,
  } = options;

  const [status, setStatus] = useState<'connecting' | 'connected' | 'disconnected'>('disconnected');
  const [messages, setMessages] = useState<WSMessage[]>([]);
  
  const wsRef = useRef<WebSocket | null>(null);
  const reconnectAttemptsRef = useRef(0);
  const reconnectTimeoutRef = useRef<number | undefined>();
  const mockIntervalRef = useRef<number | undefined>();

  const connect = useCallback(() => {
    if (MOCK_MODE) {
      setStatus('connected');
      startMockMessages();
      return;
    }

    try {
      const ws = new WebSocket(`${WS_URL}/v1/logs`);
      wsRef.current = ws;

      ws.onopen = () => {
        console.log('WebSocket connected');
        setStatus('connected');
        reconnectAttemptsRef.current = 0;
      };

      ws.onmessage = (event) => {
        try {
          const message: WSMessage = JSON.parse(event.data);
          setMessages((prev: WSMessage[]) => [...prev.slice(-99), message]); // Keep last 100 messages
          onMessage?.(message);
        } catch (error) {
          console.error('Failed to parse WebSocket message:', error);
        }
      };

      ws.onerror = (error) => {
        console.error('WebSocket error:', error);
      };

      ws.onclose = () => {
        console.log('WebSocket disconnected');
        setStatus('disconnected');
        wsRef.current = null;

        // Attempt reconnection with exponential backoff
        if (reconnectAttemptsRef.current < maxReconnectAttempts) {
          const timeout = Math.min(
            reconnectInterval * Math.pow(2, reconnectAttemptsRef.current),
            30000
          );
          
          console.log(`Reconnecting in ${timeout}ms (attempt ${reconnectAttemptsRef.current + 1})`);
          reconnectAttemptsRef.current++;
          
          reconnectTimeoutRef.current = setTimeout(() => {
            setStatus('connecting');
            connect();
          }, timeout);
        }
      };
    } catch (error) {
      console.error('Failed to create WebSocket:', error);
      setStatus('disconnected');
    }
  }, [reconnectInterval, maxReconnectAttempts, onMessage]);

  const startMockMessages = () => {
    let counter = 0;
    mockIntervalRef.current = setInterval(() => {
      const mockMessages: WSMessage[] = [
        {
          type: 'stt.partial',
          payload: { id: `${Date.now()}`, text: 'open visual studio code', partial: true, timestamp: Date.now() },
          timestamp: Date.now(),
        },
        {
          type: 'stt.final',
          payload: { id: `${Date.now()}`, text: 'open visual studio code', partial: false, timestamp: Date.now() },
          timestamp: Date.now(),
        },
        {
          type: 'intent.preview',
          payload: {
            intent: 'open_app',
            confidence: 0.95,
            slots: { app_name: 'Visual Studio Code' },
            requiresApproval: false,
            timestamp: Date.now(),
          },
          timestamp: Date.now(),
        },
        {
          type: 'action.result',
          payload: {
            ok: true,
            data: { message: 'App opened successfully' },
            requestId: `req-${Date.now()}`,
            timestamp: Date.now(),
          },
          timestamp: Date.now(),
        },
      ];

      const message = mockMessages[counter % mockMessages.length];
      setMessages((prev: WSMessage[]) => [...prev.slice(-99), message]);
      onMessage?.(message);
      counter++;
    }, 3000);
  };

  const disconnect = useCallback(() => {
    if (reconnectTimeoutRef.current) {
      clearTimeout(reconnectTimeoutRef.current);
    }
    if (mockIntervalRef.current) {
      clearInterval(mockIntervalRef.current);
    }
    if (wsRef.current) {
      wsRef.current.close();
      wsRef.current = null;
    }
    setStatus('disconnected');
  }, []);

  useEffect(() => {
    connect();
    return () => disconnect();
  }, [connect, disconnect]);

  return {
    status,
    messages,
    connect,
    disconnect,
  };
}
